ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/TopicStream.java

```diff
@@ -52,96 +60,196 @@ private TopicStream() {
 
     private final KafkaAvroDeserializer avroDeserializer;
     private final String topicName;
-    private final DateFormat dateFormat =
-        SimpleDateFormat.getDateTimeInstance(3, 1, Locale.getDefault());
+    private final DateFormat dateFormat;
+
+    private Optional<Formatter> keyFormatter = Optional.empty();
+    private Optional<Formatter> valueFormatter = Optional.empty();
 
-    private Formatter formatter;
+    public RecordFormatter(
+        final SchemaRegistryClient schemaRegistryClient,
+        final String topicName
+    ) {
+      this(
+          schemaRegistryClient,
+          topicName,
+          SimpleDateFormat.getDateTimeInstance(3, 1, Locale.getDefault())
+      );
+    }
 
-    public RecordFormatter(final SchemaRegistryClient schemaRegistryClient,
-                           final String topicName) {
-      this.topicName = Objects.requireNonNull(topicName, "topicName");
+    @VisibleForTesting
+    RecordFormatter(
+        final SchemaRegistryClient schemaRegistryClient,
+        final String topicName,
+        final DateFormat dateFormat
+    ) {
+      this.topicName = requireNonNull(topicName, "topicName");
       this.avroDeserializer = new KafkaAvroDeserializer(schemaRegistryClient);
+      this.dateFormat = requireNonNull(dateFormat, "dateFormat");
     }
 
-    public List<String> format(final ConsumerRecords<String, Bytes> records) {
-      return StreamSupport
-          .stream(records.records(topicName).spliterator(), false)
-          .filter(Objects::nonNull)
-          .filter(r -> r.value() != null)
-          .filter(r -> r.value().get() != null)
-          .filter(r -> r.value().get().length != 0)
-          .map((record) -> {
-            if (formatter == null) {
-              formatter = getFormatter(record);
-            }
-            try {
-              return formatter.print(record);
-            } catch (IOException e) {
-              log.warn("Exception formatting record", e);
-              return null;
-            }
-          })
-          .filter(Objects::nonNull)
+    public List<Supplier<String>> format(final Iterable<ConsumerRecord<Bytes, Bytes>> records) {
+      if (!keyFormatter.isPresent()) {
+        keyFormatter = getKeyFormatter(records);
+      }
+
+      if (!valueFormatter.isPresent()) {
+        valueFormatter = getValueFormatter(records);
+      }
+
+      return StreamSupport.stream(records.spliterator(), false)
+          .map(this::delayedFormat)
           .collect(Collectors.toList());
     }
 
-    public Format getFormat() {
-      return formatter == null ? Format.UNDEFINED : formatter.getFormat();
+    @SuppressWarnings("OptionalGetWithoutIsPresent") // will not be empty if needed
```

Comment:
nit(personal preference): for internal variables I actually prefer nulls to Optionals as we're not exposing it anywhere and then we don't need to keep calling get everywhere when we know it'll be not-null at that point

Reply:
Next PR removes these anyway.